lib/extras/codec_jpg.cc

```diff
@@ -408,27 +406,25 @@ Status EncodeWithLibJpeg(const ImageBundle* ib, const CodecInOut* io,
   if (!ib->IsSRGB()) {
     WriteICCProfile(&cinfo, ib->c_current().ICC());
   }
-  WriteExif(&cinfo, io->blobs.exif);
+  if (!io->blobs.exif.empty()) {
+    PaddedBytes exif(io->blobs.exif);
+    ResetExifOrientation(exif);
+    WriteExif(&cinfo, exif);
+  }
   if (cinfo.input_components > 3 || cinfo.input_components < 0)
     return JXL_FAILURE("invalid numbers of components");
 
-  std::unique_ptr<JSAMPLE[]> row(
-      new JSAMPLE[cinfo.input_components * cinfo.image_width]);
-  for (size_t y = 0; y < ib->ysize(); ++y) {
-    const float* const JXL_RESTRICT input_row[3] = {
-        ib->color().ConstPlaneRow(0, y), ib->color().ConstPlaneRow(1, y),
-        ib->color().ConstPlaneRow(2, y)};
-    for (size_t x = 0; x < ib->xsize(); ++x) {
-      for (size_t c = 0; c < static_cast<size_t>(cinfo.input_components); ++c) {
-        JXL_RETURN_IF_ERROR(c < 3);
-        row[cinfo.input_components * x + c] = static_cast<JSAMPLE>(
-            std::max(std::min(kJPEGSampleMultiplier * input_row[c][x] + .5f,
-                              kJPEGSampleMax),
-                     kJPEGSampleMin));
-      }
-    }
-    JSAMPROW rows[] = {row.get()};
-    jpeg_write_scanlines(&cinfo, rows, 1);
+  size_t stride = ib->oriented_xsize() * cinfo.input_components;
+  PaddedBytes raw_bytes(stride * ib->oriented_ysize());
+  JXL_RETURN_IF_ERROR(ConvertToExternal(
+      *ib, 8, /*float_out=*/false, cinfo.input_components, JXL_BIG_ENDIAN,
+      stride, nullptr, raw_bytes.data(), raw_bytes.size(),
+      /*out_callback=*/nullptr,
+      /*out_opaque=*/nullptr, ib->metadata()->GetOrientation()));
```

Comment:
I believe this drops support for 12-bit libjpeg, is that correct? Are we fine with that? (I assume we are, just making sure 😁)

Reply:
Ah, I didn't think about that possibility. Does such a libjpeg have JSAMPLE/JSAMPROW defined as a 16-bit type then? This code could be adapted to do both 8 and 16 bit, I just didn't realize that JSAMPLE can be something else than uint8_t. I was just trying to leverage the orientation being done by ConvertToExternal, simplifying the code a bit in the process.

In my opinion, if you care enough about precision of your output to make 12-bit libjpeg worthwhile, then most likely you'll prefer your output as a 16-bit png and not as a jpeg. But I do not object to supporting 12-bit libjpeg, I suppose it's just a matter of multiplying `stride` and bitdepth `8` by `sizeof(JSAMPLE)`, right?

Reply:
Haven't tested it with an actual 12-bit libjpeg, but I think I modified it so it should work when compiled against a 12-bit libjpeg.